src/main/java/org/truffleruby/core/cast/ToSNode.java

```diff
@@ -33,20 +33,27 @@ protected DynamicObject kernelToS(Object object) {
         return kernelToSNode.executeToS(object);
     }
 
+    private CallDispatchHeadNode getCallToSNode() {
+        if (callToSNode == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            callToSNode = insert(CallDispatchHeadNode.createOnSelf());
+        }
+        return callToSNode;
+    }
+
     @Specialization(guards = "isRubyString(string)")
     public DynamicObject toS(DynamicObject string) {
         return string;
     }
 
     @Specialization(guards = "!isRubyString(object)")
     public DynamicObject toSFallback(VirtualFrame frame, Object object) {
-        final Object value = callToSNode.call(frame, object, "to_s");
+        final Object value = getCallToSNode().call(frame, object, "to_s");
```

Comment:
Could it be created in `@Cached` or is that worse for footprint?

Reply:
fixed